File: app/src/main/java/com/noodle/sample/ui/BookListFragment.java
package com.noodle.sample.ui;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.noodle.sample.NoodleApp;
import com.noodle.sample.R;
import com.noodle.sample.data.Book;
import com.noodle.sample.data.BookManager;

import java.util.ArrayList;
import java.util.List;

public class BookListFragment extends Fragment {

  BookManager bookManager;
  BookListAdapter adapter;

  RecyclerView recyclerView;

  public BookListFragment() {
    setHasOptionsMenu(true);
  }

  @Override
  public void onCreate(@Nullable final Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    bookManager = ((NoodleApp) getActivity().getApplication()).getBookManager();
  }

  @Override
  public void onCreateOptionsMenu(final Menu menu, final MenuInflater inflater) {
    inflater.inflate(R.menu.fragment_book_list, menu);
    final SearchView searchView = (SearchView) menu.findItem(R.id.search).getActionView();
    searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
      @Override
      public boolean onQueryTextSubmit(final String query) {
        filterItems(query);
        return false;
      }

      @Override
      public boolean onQueryTextChange(final String newText) {
        filterItems(newText);
        return false;
      }
    });
  }

  @Nullable
  @Override
  public View onCreateView(final LayoutInflater inflater, @Nullable final ViewGroup container, @Nullable final Bundle savedInstanceState) {
    return inflater.inflate(R.layout.fragment_book_list, container, false);
  }

  @Override
  public void onViewCreated(final View view, @Nullable final Bundle savedInstanceState) {
    recyclerView = ((RecyclerView) view.findViewById(R.id.recyclerView));
    recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));

    adapter = new BookListAdapter(bookManager.getBooks());
    recyclerView.setAdapter(adapter);

    bookManager.setListener(new BookManager.Listener() {
      @Override
      public void onBooksChanged() {
        adapter.setItems(bookManager.getBooks());
      }
    });
  }

  @Override
  public boolean onOptionsItemSelected(final MenuItem item) {
    if (item.getItemId() == R.id.clear) {
      onClear();
    }
    return super.onOptionsItemSelected(item);
  }

  private void filterItems(final String query) {
    if (TextUtils.isEmpty(query)) {
      adapter.setItems(bookManager.getBooks());
      return;
    }

    adapter.setItems(bookManager.search(query));
  }

  void onClear() {
    new AlertDialog.Builder(getActivity())
        .setTitle(R.string.remove_all_title)
        .setMessage(R.string.remove_all_message)
        .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
          @Override
          public void onClick(final DialogInterface dialog, final int which) {
            bookManager.clear();
          }
        })
        .setNegativeButton(R.string.no, null)
        .show();
  }

  class BookListAdapter extends RecyclerView.Adapter<BookViewHolder> {

    List<Book> bookList = new ArrayList<>();
    LayoutInflater layoutInflater = getActivity().getLayoutInflater();

    BookListAdapter(final List<Book> bookList) {
      this.bookList = bookList;
    }

    @Override
    public BookViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
      final View view = layoutInflater.inflate(R.layout.item_book, parent, false);

      return new BookViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final BookViewHolder holder, final int position) {
      final Book book = bookList.get(position);

      holder.bookTitle.setText(book.title);
      holder.bookAuthor.setText(book.authorName);

      holder.itemView.setOnClickListener(new View.OnClickListener() {
        @Override
        public void onClick(final View v) {
          final EditBookFragment editFragment = new EditBookFragment();
          final Bundle args = new Bundle();
          args.putLong(EditBookFragment.ARG_BOOK_ID, book.id);
          editFragment.setArguments(args);

          editFragment.show(getFragmentManager(), "edit_book");
        }
      });
    }

    @Override
    public int getItemCount() {
      return bookList.size();
    }

    void setItems(final List<Book> items) {
      this.bookList = new ArrayList<>(items);
      notifyDataSetChanged();
    }
  }

  class BookViewHolder extends RecyclerView.ViewHolder {

    TextView bookTitle;
    TextView bookAuthor;

    BookViewHolder(final View itemView) {
      super(itemView);
      bookTitle = ((TextView) itemView.findViewById(R.id.bookTitle));
      bookAuthor = ((TextView) itemView.findViewById(R.id.bookAuthor));
    }

  }
}

File: app/src/main/java/com/noodle/sample/ui/BookListFragment.kt
package com.noodle.sample.ui

import android.os.Bundle
import android.text.TextUtils
import android.view.*
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.noodle.sample.NoodleApp
import com.noodle.sample.R
import com.noodle.sample.data.Book
import com.noodle.sample.data.BookManager
import java.util.*

class BookListFragment : Fragment() {

    private lateinit var bookManager: BookManager
    private lateinit var adapter: BookListAdapter

    private lateinit var recyclerView: RecyclerView

    init {
        setHasOptionsMenu(true)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bookManager = (requireActivity().application as NoodleApp).bookManager
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.fragment_book_list, menu)
        val searchView = menu!!.findItem(R.id.search).actionView as SearchView
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                filterItems(query)
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                filterItems(newText)
                return false
            }
        })
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.fragment_book_list, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        recyclerView = view.findViewById(R.id.recyclerView) as RecyclerView
        recyclerView.layoutManager = LinearLayoutManager(activity)

        adapter = BookListAdapter(bookManager.books)
        recyclerView.adapter = adapter

        bookManager.setListener(object : BookManager.Listener {
            override fun onBooksChanged() {
                adapter.setItems(bookManager.books)
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == R.id.clear) {
            onClear()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun filterItems(query: String) {
        if (TextUtils.isEmpty(query)) {
            adapter.setItems(bookManager.books)
            return
        }

        adapter.setItems(bookManager.search(query))
    }

    private fun onClear() {
        AlertDialog.Builder(requireContext())
                .setTitle(R.string.remove_all_title)
                .setMessage(R.string.remove_all_message)
                .setPositiveButton(R.string.yes) { _, _ -> bookManager.clear() }
                .setNegativeButton(R.string.no, null)
                .show()
    }

    internal inner class BookListAdapter(
            private var bookList: List<Book>
    ) : RecyclerView.Adapter<BookViewHolder>() {

        private val layoutInflater: LayoutInflater = LayoutInflater.from(requireContext())


        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BookViewHolder {
            val view = layoutInflater.inflate(R.layout.item_book, parent, false)

            return BookViewHolder(view)
        }

        override fun onBindViewHolder(holder: BookViewHolder, position: Int) {
            val book = bookList[position]

            holder.bookTitle.text = book.title
            holder.bookAuthor.text = book.authorName

            holder.itemView.setOnClickListener {
                val editFragment = EditBookFragment()
                val args = Bundle()
                args.putLong(EditBookFragment.ARG_BOOK_ID, book.id)
                editFragment.arguments = args

                editFragment.show(fragmentManager, "edit_book")
            }
        }

        override fun getItemCount(): Int {
            return bookList.size
        }

        fun setItems(items: List<Book>) {
            this.bookList = ArrayList(items)
            notifyDataSetChanged()
        }
    }

    internal inner class BookViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var bookTitle: TextView = itemView.findViewById(R.id.bookTitle) as TextView
        var bookAuthor: TextView = itemView.findViewById(R.id.bookAuthor) as TextView

    }
}
